File: chapter1/Quantity/Quantity.java
class Quantity{
    public final int value;
    private static final int MIN = 1;
    private static final int MAX = 100;
    
    public Quantity(int value){
        if(value < MIN){
            throw new IllegalArgumentException("value is below " + MIN);
        }
        if(value > MAX){
            throw new IllegalArgumentException("value is above " + MAX);
        }
        this.value = value;
    }

    public Quantity add(Quantity other){
        if(!canAdd(other)){
            throw new IllegalArgumentException("added over " + MAX);
        }
        int added = addValue(other);
        return new Quantity(added);
    }

    private boolean canAdd(Quantity other){
        int added = addValue(other);
        return added <= MAX;
    }

    private int addValue(Quantity other){
        return this.value + other.value;
    }

    public static void main(String[] args){
        final Quantity before = new Quantity(10);
        final Quantity toAdd = new Quantity(5);
        final Quantity result = before.add(toAdd);
        System.out.println("result value = " + result.value);
    }
}

File: chapter1/quantity-go/models/quantity.go
package models

import "fmt"

const (
	min = 1
	max = 100
)

type Quantity struct {
	value int
}

func NewQuantity(value int) (*Quantity, error) {
	if value < min {
		return nil, fmt.Errorf("value is below %d", min)
	}
	if value > max {
		return nil, fmt.Errorf("value is above %d", max)
	}
	return &Quantity{value: value}, nil
}

func (q *Quantity) Add(other *Quantity) (*Quantity, error) {
	if !q.canAdd(other) {
		return nil, fmt.Errorf("added over %d", max)
	}
	added := q.addValue(other)
	return &Quantity{value: added}, nil
}

func (q *Quantity) Value() int {
	return q.value
}

func (q *Quantity) canAdd(other *Quantity) bool {
	added := q.addValue(other)
	return added <= max
}

func (q *Quantity) addValue(other *Quantity) int {
	return q.value + other.Value()
}
